datastore/sync_entity.go

```diff
@@ -263,6 +293,164 @@ func (dynamo *Dynamo) InsertSyncEntitiesWithServerTags(entities []*SyncEntity) e
 	return nil
 }
 
+// DisableSyncChain marks a chain as disabled so no further updates or commits can happen
+func (dynamo *Dynamo) DisableSyncChain(clientID string) error {
+	now := aws.Int64(utils.UnixMilli(time.Now()))
+	disabledMarker := SyncEntity{
+		ClientID:      clientID,
+		ID:            clientID,
+		Deleted:       aws.Bool(true),
+		Version:       aws.Int64(0),
+		Mtime:         now,
+		Ctime:         now,
+		Specifics:     []byte{},
+		DataType:      aws.Int(3),
+		Folder:        aws.Bool(false),
+		DataTypeMtime: aws.String(fmt.Sprintf("3#%d", now)),
+	}
+
+	av, err := dynamodbattribute.MarshalMap(disabledMarker)
+	if err != nil {
+		return fmt.Errorf("error marshalling sync item to insert sync entity: %w", err)
+	}
+
+	markerInput := &dynamodb.PutItemInput{
+		Item:      av,
+		TableName: aws.String(Table),
+	}
+
+	_, err = dynamo.PutItem(markerInput)
+	if err != nil {
+		return fmt.Errorf("error calling PutItem to insert sync item: %w", err)
+	}
+
+	return nil
+}
+
+// ClearServerData deletes all items for a given clientID
+func (dynamo *Dynamo) ClearServerData(clientID string) ([]SyncEntity, error) {
+	syncEntities := []SyncEntity{}
+	pkb := expression.Key(pk)
+	pkv := expression.Value(clientID)
+	keyCond := expression.KeyEqual(pkb, pkv)
+	exprs := expression.NewBuilder().WithKeyCondition(keyCond)
+	expr, err := exprs.Build()
+	if err != nil {
+		return syncEntities, fmt.Errorf("error building expression to get updates: %w", err)
+	}
+
+	input := &dynamodb.QueryInput{
+		ExpressionAttributeNames:  expr.Names(),
+		ExpressionAttributeValues: expr.Values(),
+		KeyConditionExpression:    expr.KeyCondition(),
+		FilterExpression:          expr.Filter(),
+		TableName:                 aws.String(Table),
+	}
+
+	out, err := dynamo.Query(input)
+	if err != nil {
+		return syncEntities, fmt.Errorf("error doing query to get updates: %w", err)
+	}
+	count := *out.Count
+
+	err = dynamodbattribute.UnmarshalListOfMaps(out.Items, &syncEntities)
+	if err != nil {
+		return syncEntities, fmt.Errorf("error unmarshalling updated sync entities: %w", err)
+	}
+
+	var i, j int64
+	for i = 0; i < count; i += maxTransactDeleteItemSize {
+		j = i + maxTransactDeleteItemSize
+		if j > count {
+			j = count
+		}
+
+		items := []*dynamodb.TransactWriteItem{}
+		for _, item := range syncEntities[i:j] {
+			if item.ClientID == item.ID && item.Deleted != nil && *item.Deleted {
+				continue
+			}
+
+			// Fail delete if race condition detected (modified time has changed).
+			if item.Version != nil {
```

Comment:
Is it an expected behavior that items processed in previous loop might already been deleted successfully?

Reply:
This is detect writes from a client the would have happened in between querying the records and calling delete on them.

Reply:
I believe the client retries if we fail here with a transient error.  Maybe @AlexeyBarabash can confirm whether this is true or not.

So, the expectation is that if in between the query we use to collect the records and one of the delete queries a client is able to update one of the records, we fail and let the client retry delete.  If we didn't we'd risk there being a record left on the server that was updated after delete.

We set the delete marker prior to deleting anything, so once we've begun deleting, no other clients will be able to sync b/c of our middleware.  And if a client is updating at the time we're deleting, the process should fail and be retried so that on retry we get all the records..

Does that make sense?  Does that seem like a reasonably correct/complete strategy for this?

Reply:
yes, the retry logic in the client is at 

https://github.com/brave/brave-core/pull/13562/files#diff-bd6e0071e9d68f6bcfeda321eab2075e23edfa3092f5e666c008535c98bc56ebR220

if there is no successful status, client retries the request to server 5 times with delay 500 msec.